Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift

```diff
@@ -469,42 +470,46 @@ extension DefaultDiagnosticConsoleFormatter {
     private func highlightSource(
         sourceLine: String,
         lineNumber: Int,
-        range: SourceRange
+        range: SourceRange,
+        _diagnostic: Diagnostic // used in a debug assertion to identify diagnostics with incorrect source ranges
     ) -> String {
         guard highlight,
               lineNumber >= range.lowerBound.line && lineNumber <= range.upperBound.line,
               !sourceLine.isEmpty
-        else { return sourceLine }
-        
-        var startColumn: Int
-        if lineNumber == range.lowerBound.line {
-            startColumn = range.lowerBound.column
-        } else {
-            startColumn = 1
+        else {
+            return sourceLine
         }
         
-        var endColumn: Int
-        if lineNumber == range.upperBound.line {
-            endColumn = range.upperBound.column
-        } else {
-            endColumn = sourceLine.count + 1
+        guard range.lowerBound.line == range.upperBound.line else {
+            // When highlighting multiple lines, highlight the full line
+            return ANSIAnnotation.sourceHighlight.applied(to: sourceLine)
         }
 
         let sourceLineUTF8 = sourceLine.utf8
-
-        let columnRange = startColumn..<endColumn
-        let startIndex = sourceLineUTF8.index(sourceLineUTF8.startIndex, offsetBy: columnRange.lowerBound - 1)
-        let endIndex = sourceLineUTF8.index(startIndex, offsetBy: columnRange.count)
-        let highlightRange = startIndex..<endIndex
         
-        let ansiAnnotation = ANSIAnnotation.sourceHighlight
-
-        var result = ""
-        result += sourceLine[sourceLine.startIndex..<highlightRange.lowerBound]
-        result += ansiAnnotation.applied(to: String(sourceLine[highlightRange]))
-        result += sourceLine[highlightRange.upperBound..<sourceLine.endIndex]
+        let highlightStart = range.lowerBound.column - 1
+        let highlightEnd = range.upperBound.column - 1
         
-        return result
+        assert(highlightStart <= sourceLineUTF8.count, {
+            """
+            Received diagnostic with incorrect source range; (\(range.lowerBound.column) ..< \(range.upperBound.column)) extends beyond the text on line \(lineNumber) (\(sourceLineUTF8.count) characters)
```

Comment:
It's interesting when I tried the example from the Swift Book, this line number seems to be incorrect:

```
warning: '`' doesn't exist at '/The-Swift-Programming-Language/LexicalStructure'
   --> TSPL.docc/ReferenceManual/LexicalStructure.md:142:145-142:146
140 | > Grammar of an identifier:
141 | >
142 + > *identifier* → *identifier-head* *identifier-characters*_?_ \
143 | > *identifier* → **`` ` ``** *identifier-head* *identifier-characters*_?_ **`` ` ``** \
144 | > *identifier* → *implicit-parameter-name* \
```

The stray backtick is on line 143, not 142. I wonder if this discrepancy has anything to do with the crash we were seeing before? Now that we have this assert here, we should take some time to find the root cause of the invalid diagnostics.

Reply:
Yes, I saw that as well. 

The other diagnostics are on the correct line so it doesn't look like a systematic off-by-one error. We'll continue to investigate that separately.